In this project, I will be exploring the SAT score data from 2001. The data represents the average SAT score for all 50 states, including average participating rate, Average math, and verbal score. The goal of this project is to do some exploratory data analysis(EDA) along with a summary using basic Python programming and data visualization. I will also try to find some other interesting insights and demonstrate it using visualization tools like Matplotlab, Plotly and Tableau.  

# Importing Modules & Data

In [8]:
## importing pandas library
import pandas as pd

## importing numpy library
import numpy as np

## importing pyplot from matplotlib inorder to do data visualization. 
from matplotlib import pyplot as plt

## these modules are imported to use plotly. 
import plotly.plotly as py
import plotly.graph_objs as go

## matplotlib inline helps showing graphs in jupyter notebook. 
%matplotlib inline

## import data as a dataframe from a local file, open it, read it and assign it to "data"
data = pd.read_csv("../assets/sat_scores.csv")
data.head()

State  Rate  Verbal  Math
0    CT    82     509   510
1    NJ    81     499   513
2    MA    79     511   515
3    NY    77     495   505
4    NH    72     520   516

## Exploratory Data Analysis (EDA)

While doing Exploratory Data Analysis, I will try to elaborate the process by approaching the process in Manually and using python's built in functions. 

In [9]:
## Take a look at the whole dataset
data.head()

State  Rate  Verbal  Math
0    CT    82     509   510
1    NJ    81     499   513
2    MA    79     511   515
3    NY    77     495   505
4    NH    72     520   516

This is a small dataset and good for any rookie learners like me who want to get their hands dirty in the process of learning data science. There are no missing values and no errors, though I have noticed that the last row for State column has the value of "All", which I assume is the mean of all 50 states. I will remove this row for a better understanding of the data and visualization. 

## EDA using built-in python functions

In [10]:
# Inplace means to commit the changes to the DataFrame
# The 'All' row is at index 51
data.drop(51, inplace=True)

# Displaying the last five rows to see if the last row was removed. 
data.tail()

State  Rate  Verbal  Math
46    UT     5     575   570
47    IA     5     593   603
48    SD     4     577   582
49    ND     4     592   599
50    MS     4     566   551

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 51 entries, 0 to 50
Data columns (total 4 columns):
State     51 non-null object
Rate      51 non-null int64
Verbal    51 non-null int64
Math      51 non-null int64
dtypes: int64(3), object(1)
memory usage: 2.0+ KB


In [12]:
## Checking to see if there are any null values that I may have missed.
data.isnull().sum()

State     0
Rate      0
Verbal    0
Math      0
dtype: int64

In [13]:
data.describe()

Rate      Verbal        Math
count  51.000000   51.000000   51.000000
mean   37.000000  532.529412  531.843137
std    27.550681   33.360667   36.287393
min     4.000000  482.000000  439.000000
25%     9.000000  501.000000  503.000000
50%    33.000000  527.000000  525.000000
75%    64.000000  562.000000  557.500000
max    82.000000  593.000000  603.000000

The 

## EDA using raw coding

In [14]:
## Converting DataFrame into list
data_list = data.values.tolist()
##data_list = data.values.T.tolist()

I am going to create a dictionary where the key would be each headers and the values would be the value of the column assigned to each of the headers. This is a good approach if we want to work with big data since dictionary is a fast and can store multiple types at a time. 

In [15]:
## first I will get the headers to that I can assigned them to keys of the dictionary. 
headers = data.columns

## Creating an alternative dictionary in order to 
another_dict = {}
another_dict[headers[1]] = [item[1] for item in data_list]
another_dict[headers[2]] = [item[2] for item in data_list]
another_dict[headers[3]] = [item[3] for item in data_list]

In [16]:
## The following code pring the Min and Max of Rate, Math and Verbal. 
for i in range(len(another_dict.keys())):
    print ("Min of", another_dict.keys()[i], "is:", min(another_dict[another_dict.keys()[i]]))
    print ("and Max of", another_dict.keys()[i], "is:", max(another_dict[another_dict.keys()[i]]))

TypeError: 'dict_keys' object is not subscriptable

In [18]:
## Writting function using only list comprehensions, no loops, to compute Standard Deviation as a part of practice. 
def std_dev(a_list):
    n = len(a_list)
    m = sum(a_list)/n
    
    std = (sum([(item-m)**2 for item in a_list])/float(n))**0.5
    return std


## Apply the function above to find the Standard Deviation for Rate, Math and Verbal. 
print ("The Standard Deviation of Rate is: {}".format(std_dev(another_dict["Rate"])))
print ("The Standard Deviation of Math is: {}".format(std_dev(another_dict["Math"])))
print ("The Standard Deviation of Verbal is: {}".format(std_dev(another_dict["Verbal"])))

The Standard Deviation of Rate is: 27.27923867605359
The Standard Deviation of Math is: 35.92987317311408
The Standard Deviation of Verbal is: 33.03198268415228


## Visualization of the data

Visualization is one of the core componemt of data science and it is personally my favorite. For this summary analysis I will use plotly. Plotly is interective, really easy to use and fun to work with.

In [19]:
x = data.Rate
data1 = [go.Histogram(x=x)];
layout = go.Layout(
    title = "Histogram #1: Distribution of Participation Rate", 
    
    xaxis = dict(
        title = "Score"
    ),
    yaxis = dict(
        title = "Frequency"
    ),
    bargap = 0.01

)
fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='basic histogram')

PlotlyRequestError: Aw, snap! You tried to use our API as the user 'mbillah', but the supplied API key doesn't match our records. You can view your API key at https://plot.ly/settings/api

You're most likely getting this message because your local credentials file isn't synced with the Plotly server you're communicating with.

Go to https://plot.ly/<language>/getting-started (e.g., https://plot.ly/python/getting-started) for more information.

Make sure that you're logged in as mbillah.

Need help? Please try searching Plotly's <a href='http://stackoverflow.com/questions/tagged/plotly'>Stack Overflow channel</a>.

In [ ]:
x = data.Math
data1 = [go.Histogram(x=x)];
layout = go.Layout(
    title = "Histogram #2: Distribution of Math", 
    
    xaxis = dict(
        title = "Score"
    ),
    yaxis = dict(
        title = "Frequency"
    ),
    bargap = 0.01

)
fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='basic histogram1')

In [ ]:
x = data.Verbal
data2 = [go.Histogram(x=x)];
layout = go.Layout(
    title = "Histogram #3: Distribution of Verbal", 
    
    xaxis = dict(
        title = "Score"
    ),
    yaxis = dict(
        title = "Frequency"
    ),
    bargap = 0.01
)
fig = go.Figure(data = data2, layout = layout)
py.iplot(fig, filename='basic histogram2')

Now let's create an overlaid histgram to compare between verbal and Math

In [ ]:
xv = data.Verbal
xm = data.Math

trace1 = go.Histogram(
    x = xv,
    name = "Verbal", 
    opacity = 0.75,
    #autobinx =False,

)
trace2 = go.Histogram(
    x = xm, 
    name = "Math", 
    opacity = 0.75,
    #autobinx = False
)

data2 = [trace1, trace2]
layout = go.Layout(
    title = "Histogram #4: Distribution of both Math and Verbal", 
    
    xaxis = dict(
        title = "Score"
    ),
    yaxis = dict(
        title = "Frequency"
    ),
    bargap = 0.01,
    #bargroupgap = 0.1,
    barmode="overlay")

fig = go.Figure(data = data2, layout = layout)
py.iplot(fig, filename = "overlaid histogram")

### Histogram Analysis

The typical assumption is for data distribution is to follow a normal distribution. However, if we look at the overlaid histogram above, we can see that none of the histograms look like a normal distribution. The only data that seems to be closest to a normal distribution is math score distribution. The distribution for the participation rate seems to be positively skewed since lots of states are not participating enough. 


In [ ]:
# Create a trace
trace = go.Scatter(
    x = data.Verbal,
    y = data.Math,
    name = "Math",
    mode = 'markers'
)

# Create a layout with title, xaxis and yaxis
layout = go.Layout(
    title = "Scatter plot #1: Math Scores vs Verbal Scores",
    xaxis = dict(title = "Verbal"),
    yaxis = dict(title = "Math")
)
data1 = [trace]

# Plot and embed in ipython notebook!
fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='basic-scatter1')

# or plot with: plot_url = py.plot(data, filename='basic-line')

In [ ]:
# Create a trace
trace = go.Scatter(
    x = data.Rate,
    y = data.Math,
    name = "Math",
    mode = 'markers'
)

# Create a layout with title, xaxis and yaxis
layout = go.Layout(
    title = "Scatter plot #2: Math Score vs Participation Rate",
    xaxis = dict(title = "Participation Rate"),
    yaxis = dict(title = "Math Score")
)
data1 = [trace]

# Plot and embed in ipython notebook!
fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='basic-scatter2')

# or plot with: plot_url = py.plot(data, filename='basic-line')

In [ ]:
# Create a trace
trace = go.Scatter(
    x = data.Rate,
    y = data.Verbal,
    name = "Verbal",
    mode = 'markers'
)

# Create a layout with title, xaxis and yaxis
layout = go.Layout(
    title = "Scatter plot #3: Verbal Score vs Participation Rate",
    xaxis = dict(title = "Participation Rate"),
    yaxis = dict(title = "Math Score")
)
data1 = [trace]

# Plot and embed in ipython notebook!
fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='basic-scatter3')

# or plot with: plot_url = py.plot(data, filename='basic-line')

In [ ]:
# Create a trace
trace = go.Scatter(
    x = data.Rate,
    y = data.Math,
    name = "Math",
    mode = 'markers'
)
trace1 = go.Scatter(
    x = data.Rate,
    y = data.Verbal,
    name = "Verbal",
    mode = "markers"

)

# Create a layout with title, xaxis and yaxis
layout = go.Layout(
    title = "Scatter plot #4: Math and Verbal Score vs Participation Rate",
    xaxis = dict(title = "Participation Rate"),
    yaxis = dict(title = "Math and Verbal Score")
)
data1 = [trace, trace1]

# Plot and embed in ipython notebook!
fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='basic-scatter4')

# or plot with: plot_url = py.plot(data, filename='basic-line')

### Scatterplot Analysis

Scatterplots are great in analyzing relationships between two variables.  We can get whole different sets of insights by looking at the data using scatterplots. For example, by looking at the scatter plots we see that There are some interesting correlations between Math and verbal scores.

States that score higher in Math also score higher in Verbal. One particularly interesting thing is that, States with low participation rate scores comparatively higher than states with high participation rate. This is likely due to some smartest students who self-select to take SAT where the colleges do not require to take SAT. Lastly, there is an outlier in the scatter plot above where the state Ohio's math sat score is 439 and the verbal score is 534. In this situation, math score is much lower than verbal score which may be normal for some students but not for a whole state.

In [ ]:
x0 = data.Rate

Rate = go.Box(y=x0, name = "Rate", boxpoints = "all")
data1 = [Rate]
layout = go.Layout(
    title = "Boxplot #1: Boxplot for Participation Rates of 50 States",
)

fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='boxplot_1')



In [ ]:
x0 = data.Math
x1 = data.Verbal

Math = go.Box(y=x0, name = "Math", boxpoints = "all")
Verbal = go.Box(y=x1, name = "Verbal", boxpoints = "all", marker = dict(color = 'rgb(10, 128, 128)'))
data1 = [Math, Verbal]
layout = go.Layout(
    title = "Boxplot #2: Boxplot for Math and Verbal Score distrubution",
)

fig = go.Figure(data = data1, layout = layout)
py.iplot(fig, filename='boxplot_2')


### Boxplot Analysis

If we look at boxplot #1 we would see that the rate data is positively skewed. Of course, we have already known that from histogram #1. However, this boxplot #1 reveals some interesting facts since we have aligned all the points along with the boxplot. If we hover our mouse over the boxplot we can see that the middle quartile or median line is not in the center of the filled rectangle which means there is more data between median and upper quartile than the median and lower quartile, in other words, positive skewness. 

Boxplot #2 also reveals some interesting insights about SAT math and verbal score distributions. first of all, we can see why the lower whisker is so far way and we already know it is because of the outlier 439. If we ignore the outlier and try to visualize the boxplot #2, we would probably see similar boxplots representing math and verbal. However, with the outlier, the boxplot shows a somewhat symmetrical shape of math scores which proves the point from histogram analysis, that math score shows a somewhat normal distribution. As for verbal distribution, the boxplot shows a small positive skewness. This is not discernable with the histogram since the histogram does not show any recognizable pattern of skewness.

# Tableau Data Visualization

![title](../assets/Sat_verbal.png)

![title](../assets/Sat_math.png)

### Tableau Visualization Analysis

The heatmap above was created using Tableau. The heatmap shows that east coast and west coast states score significantly lower than the middle states in both math and verbal even though the middle states have lower participation rate. It seems like there is a negative correlation between rate and math or rate and verbal. 